velox/functions/prestosql/aggregates/ReduceAgg.cpp

```diff
@@ -502,6 +387,393 @@ class ReduceAgg : public exec::Aggregate {
     return lambda;
   }
 
+  // Combines 'numRows' states from 'input' with an optional 'accumulator' state
+  // and writes the result back to the accumulator.
+  void addSingleGroup(VectorPtr& input, vector_size_t numRows, char* group) {
+    // Combine 'input' states and an optional 'accumulator' state into one.
+    vector_size_t numCombined = numRows;
+    auto& accumulator = *value<ReduceAggAccumulator>(group);
+    if (accumulator.hasValue()) {
+      prepareToAppendOne(input, numRows);
+      accumulator.read(input, numRows);
+      ++numCombined;
+    }
+
+    if (numCombined > 1) {
+      input = combine(input, numCombined);
+    }
+
+    auto tracker = trackRowSize(group);
+    accumulator.write(input.get(), 0, allocator_);
+  }
+
+  BufferPtr collectNonNullRows(
+      const VectorPtr& input,
+      const SelectivityVector& rows) {
+    BufferPtr indices = allocateIndices(rows.size(), allocator_->pool());
+    auto* rawIndices = indices->asMutable<vector_size_t>();
+
+    vector_size_t i = 0;
+    rows.applyToSelected([&](auto row) {
+      if (!input->isNullAt(row)) {
+        rawIndices[i++] = row;
+      }
+    });
+
+    indices->setSize(sizeof(vector_size_t) * i);
+    return indices;
+  }
+
+  void checkStatesNotNull(const VectorPtr& states, vector_size_t size) {
+    if (states->mayHaveNulls()) {
+      for (auto i = 0; i < size; ++i) {
```

Comment:
is there a better SIMD-based version of this that can be used here instead of the explicit loop?

Reply:
This didn't show up high on the profile. To optimize we would need to first decode the vector, then generate combined nulls buffer, then count null bits. Not too hard, but more code / bugs / complexity and no performance win given all other costs.